### ISSUE TITLE
Remove double pointers from generated interfaces

### ISSUE THREAD

**Issue by garth-wells** [FEniCS/ffcx#58]
Double pointers limit possibility of using OpenCL, etc, and  are not great for readability.

Would be better to send in a flat array, with an index array for each coefficient field.

**Comment by michalhabera** [comment]
Can you be more specific/have some examples? 

I agree, that double pointers are not efficient as far as data is scattered in some big array, especially when elements are further than cache line size. The only efficient way I can think of is passing pointer to contigous data aligned somewhere outside the kernel. I have experimented with this recently.

**Comment by garth-wells** [comment]
The `tabulate_tensor` function:
```
void tabulate_tensor(ufc_scalar_t* restrict A, const ufc_scalar_t* const* w,
                                  const double* restrict coordinate_dofs,
                                  int cell_orientation);
```

**Comment by chrisrichardson** [comment]
I already used a flat array in `Form` (see `_w` in `Form.h`). The issue is really in `ffc` which needs to know where/how to index into the flat array. In principle, it knows this already, so we don't need to pass the "index array" into `tabulate_tensor()`.

**Comment by garth-wells** [comment]
Good point. Instead of 
```
double foo = w[field][component];
```
it should be simple to do 
```
double foo = w[field_offset + component];
```

**Comment by michalhabera** [comment]
@chrisrichardson Fixed with https://github.com/FEniCS/ffcx/pull/77 now?

**Comment by chrisrichardson** [comment]
yes fixed by #77
